substrate/client/rpc-spec-v2/src/common/connections.rs

```diff
@@ -0,0 +1,228 @@
+	}
+
+	/// Gives back the reserved space before the token is registered.
+	///
+	/// # Note
+	///
+	/// This may happen if the pending subscription cannot be accepted (unlikely).
+	fn unreserve_token(&self, connection_id: ConnectionId) {
+		let mut data = self.data.lock();
+
+		let entry = data.entry(connection_id).or_insert_with(ConnectionData::default);
+		entry.num_tokens = entry.num_tokens.saturating_sub(1);
+	}
+
+	/// Register a token for the given connection.
+	///
+	/// Users should call [`Self::reserve_token`] before calling this method.
+	fn register_token(&self, connection_id: ConnectionId, token: String) {
+		let mut data = self.data.lock();
+
+		let entry = data.entry(connection_id).or_insert_with(ConnectionData::default);
+		// Should be already checked `Self::reserve_token`.
+		if entry.tokens.len() >= self.capacity {
+			return;
+		}
+
+		entry.tokens.insert(token);
+	}
+
+	/// Unregister a token for the given connection.
+	fn unregister_token(&self, connection_id: ConnectionId, token: &str) {
+		let mut data = self.data.lock();
+		if let Some(connection_data) = data.get_mut(&connection_id) {
+			connection_data.tokens.remove(token);
+			connection_data.num_tokens = connection_data.num_tokens.saturating_sub(1);
+
+			if connection_data.num_tokens == 0 {
+				data.remove(&connection_id);
+			}
+		}
+	}
+
+	/// Check if the given connection contains the given token.
+	pub fn contains_token(&self, connection_id: ConnectionId, token: &str) -> bool {
+		let data = self.data.lock();
+		data.get(&connection_id)
+			.map(|connection_data| connection_data.tokens.contains(token))
+			.unwrap_or(false)
+	}
+}
+
+/// RAII wrapper that ensures the reserved space is given back if the object is
+/// dropped before the token is registered.
+pub struct ReservedConnectionToken {
+	connection_id: ConnectionId,
+	rpc_connections: Option<RpcConnections>,
+}
+
+impl ReservedConnectionToken {
+	/// Register the token for the given connection.
+	pub fn register(mut self, token: String) -> RegisteredConnectionToken {
+		let rpc_connections = self
+			.rpc_connections
+			.take()
+			.expect("Always constructed with rpc connections; qed");
+
+		rpc_connections.register_token(self.connection_id, token.clone());
+		RegisteredConnectionToken { connection_id: self.connection_id, token, rpc_connections }
+	}
+}
+
+impl Drop for ReservedConnectionToken {
+	fn drop(&mut self) {
+		if let Some(rpc_connections) = self.rpc_connections.take() {
+			rpc_connections.unreserve_token(self.connection_id);
+		}
+	}
+}
+
+/// RAII wrapper that ensures the token is unregistered if the object is dropped.
+pub struct RegisteredConnectionToken {
+	connection_id: ConnectionId,
+	token: String,
+	rpc_connections: RpcConnections,
+}
+
+impl Drop for RegisteredConnectionToken {
+	fn drop(&mut self) {
+		self.rpc_connections.unregister_token(self.connection_id, &self.token);
+	}
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Would useful with another test "is_empty" that connection IDs are cleaned from the HashMaps etc.

The code looks go to me otherwise

Reply:
Added a few extra checks to ensure we clean-up everything now, thanks!